lib/zstd_compress.c

```diff
@@ -1504,118 +1689,195 @@ static size_t ZSTD_BtFindBestMatch_selectMLS_extDict (
 }
 
 
-/* ***********************
-*  Hash Chain
-*************************/
-
-#define NEXT_IN_CHAIN(d, mask)   chainTable[(d) & mask]
 
-/* Update chains up to ip (excluded)
-   Assumption : always within prefix (ie. not within extDict) */
+/* *******************************
+*  Greedy parser
+*********************************/
 FORCE_INLINE
-U32 ZSTD_insertAndFindFirstIndex (ZSTD_CCtx* zc, const BYTE* ip, U32 mls)
+void ZSTD_compressBlock_greedy_generic(ZSTD_CCtx* ctx,
+                                     const void* src, size_t srcSize)
 {
-    U32* const hashTable  = zc->hashTable;
-    const U32 hashLog = zc->params.cParams.hashLog;
-    U32* const chainTable = zc->chainTable;
-    const U32 chainMask = (1 << zc->params.cParams.chainLog) - 1;
-    const BYTE* const base = zc->base;
-    const U32 target = (U32)(ip - base);
-    U32 idx = zc->nextToUpdate;
-
-    while(idx < target) {
-        size_t const h = ZSTD_hashPtr(base+idx, hashLog, mls);
-        NEXT_IN_CHAIN(idx, chainMask) = hashTable[h];
-        hashTable[h] = idx;
-        idx++;
-    }
+    seqStore_t* seqStorePtr = &(ctx->seqStore);
+    const BYTE* const istart = (const BYTE*)src;
+    const BYTE* ip = istart;
+    const BYTE* anchor = istart;
+    const BYTE* const iend = istart + srcSize;
+    const BYTE* const ilimit = iend - 8;
+    const BYTE* const base = ctx->base + ctx->dictLimit;
 
-    zc->nextToUpdate = target;
-    return hashTable[ZSTD_hashPtr(ip, hashLog, mls)];
-}
+    const U32 maxSearches = 1 << ctx->params.cParams.searchLog;
+    const U32 mls = ctx->params.cParams.searchLength;
 
+    /* init */
+    U32 rep[ZSTD_REP_INIT];
+    for (U32 i=0; i<ZSTD_REP_INIT; i++)
```

Comment:
While it's a good idea to converge towards C90,
hence to convert all `//` into `/* ... */`, _at least in the library itself_,
it's not a mandatory target either.

Reason is, strict C90 is very limited, especially regarding 64-bits types.

But the world is not divided between "strict c90" and "strict c99".
This would be too easy.
In fact, most compilers which are not completely c99 do support a subset of it,
among which the `//` comment tag is the most common.

It's pretty hard to find a compiler which doesn't support `//`,
but unfortunately it's very common for non-c99 compilers to not support `if (U32 i=0; ...`
Hence I put more pressure on avoiding this construction.

Bottom line : yes, try to stick to c90 whenever possible, including regarding comments tags.